src/ekstat.erl

```diff
@@ -1,65 +1,162 @@
+    {Statistic::string(), Value::string()}.
+-type kstat_row() ::
+    {Class::string(), Module::string(), Instance::integer(),
+        Name::string(), [kstat_statistic()]}.
 
 -export_type([kstat_handle/0, kstat_pattern/0]).
 
 %%%===================================================================
 %%% API functions
 %%%===================================================================
 
--spec open() -> {ok, Handle::kstat_handle()} | kstat_error().
+-spec open() ->
+    {ok, Handle::kstat_handle()} | kstat_error().
 open() ->
+    ?ASYNC_NIF_CALL(fun open_nif/1, []).
+
+-spec close(Handle) ->
+    ok | kstat_error()
+    when
+        Handle :: binary().
+close(Handle) ->
+    ?ASYNC_NIF_CALL(fun close_nif/2, [Handle]).
+
+-spec clear(Handle) ->
+    {ok, NumInstances::integer()} | kstat_error()
+    when
+        Handle :: binary().
+clear(Handle) ->
+    ?ASYNC_NIF_CALL(fun clear_nif/2, [Handle]).
+
+-spec update(Handle) ->
+    {ok, NumInstances::integer()} | kstat_error()
+    when
+        Handle :: binary().
+update(Handle) ->
+    ?ASYNC_NIF_CALL(fun update_nif/2, [Handle]).
+
+-spec read(Handle) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary().
+read(Handle) ->
+    ?ASYNC_NIF_CALL(fun read_nif/2, [Handle]).
+
+-spec read(Handle, Class) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary(),
+        Class :: kstat_pattern().
+read(Handle, Class) ->
+    ?ASYNC_NIF_CALL(fun read_nif/3, [Handle, Class]).
+
+-spec read(Handle, Class, Module) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary(),
+        Class :: kstat_pattern(),
+        Module :: kstat_pattern().
+read(Handle, Class, Module) ->
+    ?ASYNC_NIF_CALL(fun read_nif/4, [Handle, Class, Module]).
+
+-spec read(Handle, Class, Module, Instance) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary(),
+        Class :: kstat_pattern(),
+        Module :: kstat_pattern(),
+        Instance :: kstat_pattern().
+read(Handle, Class, Module, Instance) ->
+    ?ASYNC_NIF_CALL(fun read_nif/5, [Handle, Class, Module, Instance]).
+
+-spec read(Handle, Class, Module, Instance, Name) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary(),
+        Class :: kstat_pattern(),
+        Module :: kstat_pattern(),
+        Instance :: kstat_pattern(),
+        Name :: kstat_pattern().
+read(Handle, Class, Module, Instance, Name) ->
+    ?ASYNC_NIF_CALL(fun read_nif/6, [Handle, Class, Module, Instance, Name]).
+
+-spec read(Handle, Class, Module, Instance, Name, Statistic) ->
+    [kstat_row()] | kstat_error()
+    when
+        Handle :: binary(),
+        Class :: kstat_pattern(),
+        Module :: kstat_pattern(),
+        Instance :: kstat_pattern(),
+        Name :: kstat_pattern(),
+        Statistic :: kstat_pattern().
+read(Handle, Class, Module, Instance, Name, Statistic) ->
+    ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic]).
```

Comment:
Line 115 is too long:     ?ASYNC_NIF_CALL(fun read_nif/7, [Handle, Class, Module, Instance, Name, Statistic])..